### PR TITLE
PCHR-2692: Allow to set Work Pattern custom hours

### PR DESCRIPTION
**Pull request by igorpavlov**
## Overview

At the moment `Number of Hours` value is calculated basing on the `Time from`, `Time to` and `Break`. This PR still leaves this pre-calculation, but also allows to set custom `Number of Hours`.

This field is auto-rounded to be divisible by 15 minutes (0.25) and properly validated for format and min/max values.

### Back-end
_uk.co.compucorp.civicrm.hrleaveandabsences/CRM/HRLeaveAndAbsences/BAO/WorkDay.php_

Removed the calculation function on the **back-end** (it still exists on the front-end).

```php
// Removed completely
private static function calculateNumberOfHours($timeFrom, $timeTo, $break) {
  // ...
  return $numberOfHours;
}
```

Make it get the value directly from the form

```php
$numberOfHours = empty($params['number_of_hours']) ? null : $params['number_of_hours'];
```

Apply validators

```php
if (!is_null($numberOfHours)) {
  if (!is_numeric($numberOfHours)) {
    // throw error
  } else if ($numberOfHours <= 0 || $numberOfHours >= 24) {
    // throw error
  } else if (strpos($numberOfHours / 0.25, '.' )) {
    // throw error
  }
}
```

### Form

Make the field editable

```php
$this->add('text', "weeks[$i][days][$j][number_of_hours]", '', ['maxlength' => 5, 'class' => 'work-day-hours']);
```

### Front-end

Added a rounding function

```js
Day.prototype._roundNumberOfHours = function() {
  var numberOfHours = parseFloat(this._numberOfHoursField.value);
  var divider = 0.25; // 15 minutes
  if (!isNaN(numberOfHours)) {
    numberOfHours = Math.round(numberOfHours / divider) * divider;
  }
  this._numberOfHoursField.value = numberOfHours;
};
// Call it in two places
_calculateNumberOfHours // <-- after calculation
_setFieldsMasks // <-- as a mask for custom editing
```

## Comments

Additionally:
- Added `no-select2` class to selectors to ensure they are rendered properly and do not rely on JavaScript anymore
- Added `crm-select` and `no-select2` classes to a selector of number of weeks to make it look consistent with other selectors
image
- Hours table row has now a border across all columns
image
- If there are form errors, the tab containing these errors will be opened automatically
err
- ⚠️ Break can now be 0
- ⚠️ Break cannot now be a negative number

---

- [x] Tests Pass

image

ℹ️ Please note, work pattern front-end does not have test suites yet.

### CODE REVIEW

File: uk.co.compucorp.civicrm.hrleaveandabsences/CRM/HRLeaveAndAbsences/BAO/WorkDay.php [hunk 1]
```diff
@@ -165,27 +157,55 @@ private static function validateWorkHours($params) {
...
     $secondsInWorkingHours = strtotime($timeTo) - strtotime($timeFrom);
     $secondsInBreak = $break * 3600;
     if($hasTimes && ($secondsInBreak >= $secondsInWorkingHours)) {
       throw new InvalidWorkDayException(
-          'Break should be less than the number of hours between Time From and Time To'
+        'Break should be less than the number of hours between Time From and Time To'
       );
     }
+
+    if (!is_null($numberOfHours)) {
```
Comment (tunbola) [inline uk.co.compucorp.civicrm.hrleaveandabsences/CRM/HRLeaveAndAbsences/BAO/WorkDay.php]: All the conditions can be in their own independent If's, no need for if/elseif, its not as if the validations depend on one another, its just a set of conditions that we need to test.
Will also reduce nesting and improve readability.

File: uk.co.compucorp.civicrm.hrleaveandabsences/CRM/HRLeaveAndAbsences/BAO/WorkDay.php [hunk 2]
```diff
@@ -165,27 +157,55 @@ private static function validateWorkHours($params) {
...
+        'Time From should be less than Time To'
       );
     }
 
+    if (!is_null($break)) {
+      if (!is_numeric($break)) {
+        throw new InvalidWorkDayException(
+          'Break should be a valid number'
+        );
+      } else if ($break < 0) {
```
Comment (tunbola) [inline uk.co.compucorp.civicrm.hrleaveandabsences/CRM/HRLeaveAndAbsences/BAO/WorkDay.php]: No need for else if here, the condition can be in its own independent if.

File: uk.co.compucorp.civicrm.hrleaveandabsences/tests/phpunit/CRM/HRLeaveAndAbsences/BAO/WorkDayTest.php [hunk 3]
```diff
@@ -72,18 +72,12 @@ public function testTimeFromTimeToAndBreakShouldBeRequiredIfItsAWorkingDay($time
 
   /**
    * @expectedException CRM_HRLeaveAndAbsences_Exception_InvalidWorkDayException
-   * @expectedExceptionMessage Time From, Time To and Break should be empty for Non Working Days and Weekends
+   * @expectedExceptionMessage Break should be a valid number
    *
-   * @dataProvider timesAndBreakEmptyDataProvider
+   * @dataProvider invalidFormatBreakDataProvider
    */
-  public function testTimeFromTimeToAndBreakShouldBeEmptyIfItsNotAWorkingDay($timeTo, $timeFrom, $break) {
```
Comment (tunbola) [inline uk.co.compucorp.civicrm.hrleaveandabsences/tests/phpunit/CRM/HRLeaveAndAbsences/BAO/WorkDayTest.php]: Don't we need this test `testTimeFromTimeToAndBreakShouldBeEmptyIfItsNotAWorkingDay` again?
We still need a test that verify the exception message: `Time From, Time To, Break and Number of Hours should be empty for Non Working Days and Weekends`

File: uk.co.compucorp.civicrm.hrleaveandabsences/tests/phpunit/CRM/HRLeaveAndAbsences/BAO/WorkDayTest.php [hunk 4]
```diff
@@ -156,16 +159,38 @@ public function testCannotHaveMoreThanOfDayOfTheWeekForEachWeek() {
...
+   * @dataProvider validNumberOfHoursDataProvider
    */
-  public function testNumberOfHoursShouldBeCalculatedFromTimesAndBreak($timeFrom, $timeTo, $break, $expectedNumberOfHours) {
-    $entity = $this->createBasicWorkDay([
-      'time_from' => $timeFrom,
-      'time_to' => $timeTo,
-      'break' => $break
-    ]);
+  public function testNumberOfHours($numberOfHours) {
+    $entity = $this->createBasicWorkDay(['number_of_hours' => $numberOfHours]);
```
Comment (tunbola) [inline uk.co.compucorp.civicrm.hrleaveandabsences/tests/phpunit/CRM/HRLeaveAndAbsences/BAO/WorkDayTest.php]: How are we sure If a Work day was created,  can we try asserting that a WorkDay instance was returned?

File: uk.co.compucorp.civicrm.hrleaveandabsences/tests/phpunit/CRM/HRLeaveAndAbsences/BAO/WorkDayTest.php [hunk 5]
```diff
@@ -156,16 +159,38 @@ public function testCannotHaveMoreThanOfDayOfTheWeekForEachWeek() {
...
-   * @dataProvider numberOfHoursDataProvider
+   * @dataProvider validNumberOfHoursDataProvider
    */
-  public function testNumberOfHoursShouldBeCalculatedFromTimesAndBreak($timeFrom, $timeTo, $break, $expectedNumberOfHours) {
-    $entity = $this->createBasicWorkDay([
-      'time_from' => $timeFrom,
-      'time_to' => $timeTo,
-      'break' => $break
-    ]);
+  public function testNumberOfHours($numberOfHours) {
```
Comment (tunbola) [inline uk.co.compucorp.civicrm.hrleaveandabsences/tests/phpunit/CRM/HRLeaveAndAbsences/BAO/WorkDayTest.php]: This test name needs a bit more description, e.g `testWorkDatsCanBeCreatedWithValidHours`.

File: uk.co.compucorp.civicrm.hrleaveandabsences/CRM/HRLeaveAndAbsences/BAO/WorkDay.php [hunk 6]
```diff
@@ -165,27 +157,55 @@ private static function validateWorkHours($params) {
...
+    if (!is_null($numberOfHours)) {
+      if (!is_numeric($numberOfHours)) {
+        throw new InvalidWorkDayException(
+          'Number of Hours should be a valid number'
+        );
+      } else if ($numberOfHours <= 0 || $numberOfHours >= 24) {
+        throw new InvalidWorkDayException(
+          'Number of Hours should be between 0 and 24'
+        );
+      } else if (strpos($numberOfHours / 0.25, '.' )) {
```
Comment (tunbola) [inline uk.co.compucorp.civicrm.hrleaveandabsences/CRM/HRLeaveAndAbsences/BAO/WorkDay.php]: What about `if(fmod($numberOfHours, 0.25) !== 0)` ?

File: uk.co.compucorp.civicrm.hrleaveandabsences/CRM/HRLeaveAndAbsences/BAO/WorkDay.php [hunk 7]
```diff
@@ -165,27 +157,55 @@ private static function validateWorkHours($params) {
...
+        throw new InvalidWorkDayException(
+          'Number of Hours should be divisible by 0.25'
```
Comment (tunbola) [inline uk.co.compucorp.civicrm.hrleaveandabsences/CRM/HRLeaveAndAbsences/BAO/WorkDay.php]: What about a more user friendly error message? 
e.g `Number of Hours are only allowed in 15 minutes blocks`

File: uk.co.compucorp.civicrm.hrleaveandabsences/tests/phpunit/CRM/HRLeaveAndAbsences/BAO/WorkDayTest.php [hunk 8]
```diff
@@ -156,16 +159,38 @@ public function testCannotHaveMoreThanOfDayOfTheWeekForEachWeek() {
...
+   */
+  public function testNumberOfHoursMustBeBetween0And24 ($numberOfHours) {
+    $entity = $this->createBasicWorkDay(['number_of_hours' => $numberOfHours]);
+  }
+
+  /**
+   * @expectedException CRM_HRLeaveAndAbsences_Exception_InvalidWorkDayException
+   * @expectedExceptionMessage Number of Hours should be divisible by 0.25
+   */
+  public function testNumberOfHoursMustBeDivisableBy15Minutes () {
```
Comment (tunbola) [inline uk.co.compucorp.civicrm.hrleaveandabsences/tests/phpunit/CRM/HRLeaveAndAbsences/BAO/WorkDayTest.php]: Maybe have a data provider for this, so we can test this with more options?

File: uk.co.compucorp.civicrm.hrleaveandabsences/CRM/HRLeaveAndAbsences/BAO/WorkDay.php [hunk 9]
```diff
@@ -165,27 +157,55 @@ private static function validateWorkHours($params) {
...
     }
+
+    if (!is_null($numberOfHours)) {
+      if (!is_numeric($numberOfHours)) {
+        throw new InvalidWorkDayException(
+          'Number of Hours should be a valid number'
+        );
+      } else if ($numberOfHours <= 0 || $numberOfHours >= 24) {
+        throw new InvalidWorkDayException(
+          'Number of Hours should be between 0 and 24'
```
Comment (tunbola) [inline uk.co.compucorp.civicrm.hrleaveandabsences/CRM/HRLeaveAndAbsences/BAO/WorkDay.php]: This means that a value of 24 will not be allowed. Is this right?

**Comment by igorpavlov** [comment]
@reneolivo backstop passed

image

It even didn't notice the border change here:

image

File: uk.co.compucorp.civicrm.hrleaveandabsences/CRM/HRLeaveAndAbsences/BAO/WorkDay.php [hunk 10]
```diff
@@ -143,13 +133,15 @@ private static function validateWorkHours($params) {
     $typeOfDay = empty($params['type']) ? null : $params['type'];
     $timeFrom = empty($params['time_from']) ? null : $params['time_from'];
     $timeTo = empty($params['time_to']) ? null : $params['time_to'];
-    $break = empty($params['break']) ? null : $params['break'];
-
-    $hasTimesOrBreak = $timeFrom || $timeTo || $break;
+    $break = isset($params['break']) && $params['break'] == '0' ? 0 :
+      (empty($params['break']) ? null : $params['break']);
```
[the remark this answers is not in the document]
Comment (igorpavlov) [reply uk.co.compucorp.civicrm.hrleaveandabsences/CRM/HRLeaveAndAbsences/BAO/WorkDay.php]: `$break = !$breakIsNotEmpty ? null : $params['break'];` that will throw a notice since `$params['break']` is not really defined in PHP in case of 0. But I **do** get your point, will improve.

File: uk.co.compucorp.civicrm.hrleaveandabsences/CRM/HRLeaveAndAbsences/BAO/WorkDay.php (hunk 2 above)
Comment (igorpavlov) [reply uk.co.compucorp.civicrm.hrleaveandabsences/CRM/HRLeaveAndAbsences/BAO/WorkDay.php]: True that, it will stop once thrown anyway.

File: uk.co.compucorp.civicrm.hrleaveandabsences/CRM/HRLeaveAndAbsences/BAO/WorkDay.php (hunk 9 above)
Comment (igorpavlov) [reply uk.co.compucorp.civicrm.hrleaveandabsences/CRM/HRLeaveAndAbsences/BAO/WorkDay.php]: To work 24 hours per day is a tiny bit inhumane, though sometimes could be pretty efficient for business :) So a good question. @davialexandre should we allow Number of Hours to be exactly 24?

File: uk.co.compucorp.civicrm.hrleaveandabsences/CRM/HRLeaveAndAbsences/BAO/WorkDay.php (hunk 10 above)
[the remark this answers is not in the document]
Comment (tunbola) [reply uk.co.compucorp.civicrm.hrleaveandabsences/CRM/HRLeaveAndAbsences/BAO/WorkDay.php]: This has already taken care of it `isset($params['break']) && $params['break'] == 0)`. It checks that break exists and is zero.

File: uk.co.compucorp.civicrm.hrleaveandabsences/CRM/HRLeaveAndAbsences/BAO/WorkDay.php (hunk 9 above)
Comment (davialexandre) [reply uk.co.compucorp.civicrm.hrleaveandabsences/CRM/HRLeaveAndAbsences/BAO/WorkDay.php]: If we allow it to be 23, I can't see a reason to not allow it to be 24.

File: uk.co.compucorp.civicrm.hrleaveandabsences/CRM/HRLeaveAndAbsences/BAO/WorkDay.php (hunk 2 above)
Comment (igorpavlov) [reply uk.co.compucorp.civicrm.hrleaveandabsences/CRM/HRLeaveAndAbsences/BAO/WorkDay.php]: Done

File: uk.co.compucorp.civicrm.hrleaveandabsences/CRM/HRLeaveAndAbsences/BAO/WorkDay.php (hunk 1 above)
Comment (igorpavlov) [reply uk.co.compucorp.civicrm.hrleaveandabsences/CRM/HRLeaveAndAbsences/BAO/WorkDay.php]: Done.

File: uk.co.compucorp.civicrm.hrleaveandabsences/CRM/HRLeaveAndAbsences/BAO/WorkDay.php (hunk 9 above)
Comment (igorpavlov) [reply uk.co.compucorp.civicrm.hrleaveandabsences/CRM/HRLeaveAndAbsences/BAO/WorkDay.php]: Done.

File: uk.co.compucorp.civicrm.hrleaveandabsences/CRM/HRLeaveAndAbsences/BAO/WorkDay.php (hunk 6 above)
Comment (igorpavlov) [reply uk.co.compucorp.civicrm.hrleaveandabsences/CRM/HRLeaveAndAbsences/BAO/WorkDay.php]: Done.

File: uk.co.compucorp.civicrm.hrleaveandabsences/CRM/HRLeaveAndAbsences/BAO/WorkDay.php (hunk 7 above)
Comment (igorpavlov) [reply uk.co.compucorp.civicrm.hrleaveandabsences/CRM/HRLeaveAndAbsences/BAO/WorkDay.php]: Done.

File: uk.co.compucorp.civicrm.hrleaveandabsences/tests/phpunit/CRM/HRLeaveAndAbsences/BAO/WorkDayTest.php (hunk 3 above)
Comment (igorpavlov) [reply uk.co.compucorp.civicrm.hrleaveandabsences/tests/phpunit/CRM/HRLeaveAndAbsences/BAO/WorkDayTest.php]: Done

File: uk.co.compucorp.civicrm.hrleaveandabsences/tests/phpunit/CRM/HRLeaveAndAbsences/BAO/WorkDayTest.php (hunk 4 above)
Comment (igorpavlov) [reply uk.co.compucorp.civicrm.hrleaveandabsences/tests/phpunit/CRM/HRLeaveAndAbsences/BAO/WorkDayTest.php]: Done

File: uk.co.compucorp.civicrm.hrleaveandabsences/tests/phpunit/CRM/HRLeaveAndAbsences/BAO/WorkDayTest.php (hunk 5 above)
Comment (igorpavlov) [reply uk.co.compucorp.civicrm.hrleaveandabsences/tests/phpunit/CRM/HRLeaveAndAbsences/BAO/WorkDayTest.php]: Done.

File: uk.co.compucorp.civicrm.hrleaveandabsences/tests/phpunit/CRM/HRLeaveAndAbsences/BAO/WorkDayTest.php (hunk 8 above)
Comment (igorpavlov) [reply uk.co.compucorp.civicrm.hrleaveandabsences/tests/phpunit/CRM/HRLeaveAndAbsences/BAO/WorkDayTest.php]: Done.

File: uk.co.compucorp.civicrm.hrleaveandabsences/CRM/HRLeaveAndAbsences/BAO/WorkDay.php (hunk 10 above)
[the remark this answers is not in the document]
Comment (igorpavlov) [reply uk.co.compucorp.civicrm.hrleaveandabsences/CRM/HRLeaveAndAbsences/BAO/WorkDay.php]: @tunbola please take my apologies, it works actually, I have no idea why it didn't earlier when I tested it.
